### ISSUE TITLE
URL rebasing from index.html

### ISSUE THREAD

**Issue by mgraszek** [jakubpawlowicz/clean-css#854]
Hi

I'm using gulp-clean-css with the version 3.4.23. I'm trying to pack all css files into one,  loading it from the index.html

`
        
        <link rel="stylesheet" type="text/css" href="style/application/bootstrap.css">
        
`

And in the gulp

`  return gulp.src('src/index.html')
    .pipe(gulpif('*.css', cleanCSS({options})))
    .pipe(uncache())
    .pipe(gulp.dest(buildDir));
`

What options should use to properly rebase the url() inside css file.
 Resources from url can be allocated in different folders, and they are copy to buildDir.
Like:
src/a/b/image.png -> buildDir/a/b/image.png

I was playing with relativeTo, target, root, rebase(true/false) and nothing seems to work.

**Comment by jakubpawlowicz** [comment]
@mgraszek couple questions

- does your CSS use relative paths to those images?
- where your output CSS will end up relative to `buildDir`?

**Comment by mgraszek** [comment]
These css files are the libraries css files. For example:
src/lib/bootstrap/css/bootstrap.css (url('../font/font1.tff')
src/lib/bootstrap/font/font1.tff

Output css file is placed inside build dir
buildDir/application.cs

**Comment by jakubpawlowicz** [comment]
Then you should be able to do it passing one of these:

```js
options = {
  target: 'src'
}
```

or

```js
options = {
  root: 'src',
  target: 'src'
}
```


`root` sets a how to resolve absolute paths while `target` sets where the output file will live. You may think that it should be `buildDir` but clean-css does not have a concept of moving files around. So by saying your output will live in `src` (even if it doesn't) and then copying all files to `buildDir` relative paths will be preserved.

Hope this helps.

**Comment by mgraszek** [comment]
It didn' work. Results are still the same.

**Comment by jakubpawlowicz** [comment]
Could you give me an example of an URL from the output file?

**Comment by mgraszek** [comment]
`
@font-face {
  font-family: 'Glyphicons Halflings';

  src: url('../fonts/glyphicons-halflings-regular.eot');
  src: url('../fonts/glyphicons-halflings-regular.eot?#iefix') format('embedded-opentype'), url('../fonts/glyphicons-halflings-regular.woff2') format('woff2'), url('../fonts/glyphicons-halflings-regular.woff') format('woff'), url('../fonts/glyphicons-halflings-regular.ttf') format('truetype'), url('../fonts/glyphicons-halflings-regular.svg#glyphicons_halflingsregular') format('svg');
}`


![image](https://cloud.githubusercontent.com/assets/11993999/21807921/38d37f50-d741-11e6-95e6-9f9edfd30e2f.png)

In the buildDir/application.css @font-face looks the same

**Comment by jakubpawlowicz** [comment]
This is odd.

I assume you don't use `rebase: false` as an option, right?
Does the same thing happen with image URLs as well?

**Comment by mgraszek** [comment]
Yes, this works the same for image URLs. 
Rebase true/false didn't change a thing.

Maybe loading from index.html is not a good way?

**Comment by jakubpawlowicz** [comment]
Thanks for more details. As far as I know loading from index.html should be the way.

I've just checked gulp-clean-css (you use this one, right - https://github.com/scniro/gulp-clean-css?) and what about these options:

```js
options = {
  relativeTo: 'src',
  target: 'src'
}
```

**Comment by mgraszek** [comment]
Yes. I'm using that gulp-clean-css.
These options didn't help.

Did you try to do something similar? In the same project i have less files and im using there gulp-less and later gulp-clean-css on them, and they work fine.  I only have problem with these files put as link in index.html.

**Comment by mgraszek** [comment]
Okay i played a little longer with this case. I was able to do minification with cleanCSS for this case, but im still not able to rebase urls but i achieve something funny.
Old urls:
src: url('../fonts/glyphicons-halflings-regular.eot');
New urls:
src: url(../fonts/glyphicons-halflings-regular.eot);

Using index.html to load css files i forgot to add userref(). Code now looks like this:
```
  return gulp.src('src/index.html')
    .pipe(useref())
    .pipe(gulpif('*.css', cleanCSS({target: 'src'})))
    .pipe(uncache())
    .pipe(gulp.dest(buildDir));
```

If i add 'root' option, im receiving errors in chrome console, that file in <fullPath> is not available

**Comment by jakubpawlowicz** [comment]
Ok, so that's something. Could you give `target` + `relativeTo` a try? If that doesn't work what are the console errors when using `root` option?

**Comment by mgraszek** [comment]
I tried that options too, they didnt work. About the root error in chrome console - for every url file i receive

``Not allowed to load local resource: file:///<fullPathToProject>/<pathFromUrl>`

**Comment by jakubpawlowicz** [comment]
Thanks for trying out other options. I have no further ideas at this point beside asking you to share an excerpt (gulpfile, index.html, and couple files from `src` folder) so I can give it a try locally.

Would you be able to share it publicly or privately (my email is in package.json)?

**Comment by mgraszek** [comment]
Okay. I was able to fix problem (or more like Jakub did it for me).
According what i got from Jakub. 'userref(), i was using, merges every files inside  ...  and this "merge" is put to cleanCSS()
The problem with url() is copying these values to merged file and do nothing with them. CleanCSS() doesn't know how to change paths to be correct and leaves them as they are.

The results we get is to put files from  to 1 css file using imports. For example:

```
        <!-- build:css application.css -->

        <link rel="stylesheet" type="text/css" href="dir1/file1.css">
        <link rel="stylesheet" type="text/css" href="dir1/file2.css">

        <!-- endbuild -->
```
and change it into

```
        <!-- build:css application.css -->

        <link rel="stylesheet" type="text/css" href="dir1/fileX.css">

        <!-- endbuild -->
```
and inside fileX.css we put @import
```
@import url(file1.css);
@import url(file2.css);

```

The important stuff is that, its working if we don't have urls like ('../dir/something/file.css'). It must be full path like ('folder1/folder2/file.css').
If we have '../' in import we will get an error.

`Error: Uncaught, unspecified "error" event. (Broken @import declaration of "../dir/something/file.css"`

**Comment by jakubpawlowicz** [comment]
Thanks @mgraszek for summing it up.
Btw, I've updated your comment to match the original paths.

As long as clean-css deals with resolving imports all URLs can be resolved correctly. 

Also if you rely on copying files manually you can point both `relativeTo` and `target` to input directory and it'll work fine, so in the original example, that would be:

```js
sourceDir = 'src';
buildDir = 'public';
...
return gulp.src('src/index.html')
  .pipe(useref())
  .pipe(gulpif('*.css', cleanCSS({ target: sourceDir, relativeTo: sourceDir })))
  .pipe(uncache())
  .pipe(gulp.dest(buildDir));
```

**Comment by sebastian-zarzycki-es** [comment]
We have the exact same problem. We include css in html but after compiling with gulp-clean-css, the paths seem to be off in the same way. I have no idea whether this is a gulp issue, but I had similar problems with grunt with the same task before. Either clean-css isn't taking some scenario into consideration, or the gulp build files incorrect - unfortunately, I cannot tell which one is it. 

The basic idea is to get all css from index.html, concatenate into one file, minify, while preserving the correct urls to resources. Apparently there's no single solution for this, that would work properly. As such, it's hard to figure out whether it's the clean-css, gulp-clean-css or gulp itself to blame.

Moving csses into another .css file with all the imports seems a bit weird/clunky. Why cannot it work when imported straight from index.html?

@jakubpawlowicz , would it be possible for you to create a simple case of gulpfile properly taking css included in index.html and concatenating/minifying them into one file, while maintaining the urls? Bootstrap css seems like a good candidate, since it references fonts as external resources. Something like:

input:
```
 lib\bootstrap\css
 lib\bootstrap\fonts
 index.html
```

output:
```
lib\bootstrap\fonts
index.html
all.css
```

With all.css having proper urls to lib\bootstrap\fonts?

I know that's asking a lot, but it would look to me like this simple usecase would be a good testing scenario for clean-css and expose and potential problems along the way.

**Comment by jakubpawlowicz** [comment]
@mgraszek if you don't mind I'll create a template loosely based on what you shared with me

@sebastian-zarzycki-es the reason it cannot import straight from index.html is (in this case) `useref` plugin which either 1) concatenates all input and pipes it to clean-css or 2) pipes files one by one as text blobs. By getting a text blob without any information where the original file was located any relative paths are impossible to resolve.

**Comment by mgraszek** [comment]
@jakubpawlowicz No problem. You can share what you want.

**Comment by jakubpawlowicz** [comment]
@mgraszek @sebastian-zarzycki-es I'm attaching a simple example to this commit. 

Assuming it's downloaded to current directory, run:

```shell
unzip gulp-rebasing-example.zip
cd gulp-rebasing-example
npm install
./node_modules/.bin/gulp build
```

And then take a look at `public/application.css` with paths correctly resolved to bootstrap and calendar assets.

Hope this helps!

[gulp-rebasing-example.zip](https://github.com/jakubpawlowicz/clean-css/files/726445/gulp-rebasing-example.zip)

**Comment by sebastian-zarzycki-es** [comment]
Thanks, this is appreciated. It's an improvement. But I still have some issues with it - mostly around using an additional css with imports. Attaching two examples to illustrate.

1 - If I move that file somewhere else in the structure, say src/resource/application.css, update the import paths to use '../lib' and update index.html accordingly, the whole build breaks with:

```
Error: Uncaught, unspecified "error" event. (Broken @import declaration of "../lib/bootstrap/css/bootstrap.css" Broken @import declaration of "../lib/bootstrap/css/bootstrap-standard.css" Broken @import declaration of "../lib/calendar/css/calendar.css")
    at emit (events.js:144:17)
```

Which "kind of" makes sense, because the import is done at /src level, and if the import css file contains ../ then it cannot reach higher than the execution location (security constraint, I guess?). This forces me to update the import paths in css as if it was in /src and not in whatever folder it actually is. Is there a way keep this consistent?

[gulp-rebasing-example-1.zip](https://github.com/jakubpawlowicz/clean-css/files/729304/gulp-rebasing-example-1.zip)

2. When I move the css definitions directly into index.html, the build outputs incorrect paths to resources in css again. Is there any way to make it work properly *while* keeping all the css imports in html?

[gulp-rebasing-example-2.zip](https://github.com/jakubpawlowicz/clean-css/files/729305/gulp-rebasing-example-2.zip)

**Comment by jakubpawlowicz** [comment]
Re 1 - that's because you need to point `relativeTo` differently, e.g. this works fine:

```js
var gulp = require('gulp');
var useref = require('gulp-useref');
var cleanCSS = require('gulp-clean-css');
var gulpif = require('gulp-if');
var uncache = require('gulp-uncache');

var targetDir = 'dist';

gulp.task('lib-resources', function () {
  return gulp.src(['src/lib/**/*', '!src/lib/**/*.css'], { base: 'src' })
    .pipe(gulp.dest(targetDir));
});

gulp.task('build', ['lib-resources'], function () {
  return gulp.src('src/index.html')
    .pipe(useref())
    .pipe(gulpif('*.css', cleanCSS({ target: 'src', relativeTo: 'src/resource' })))
    .pipe(uncache())
    .pipe(gulp.dest(targetDir));
});
```

Re 2 - this is not up to clean-css as it receives a piped file from gulp, namely useref plugin. I'm pretty sure you can patch useref to turn all `<link...` statements into `@import` rules, so do the current manual `application.css` step in-memory.

**Comment by sebastian-zarzycki-es** [comment]
I'm not sure I follow. Correct me if I'm wrong here.

In both cases, useref() is providing the files. I think that the difference is that with one file (one css with import), there's just plain include with known paths, so that we have one "entry point", if you will - source, target and relativeTo - we can hardcode it for cleanCSS. In the second one, there are multiple files with different locations, which means that single relativeTo won't work and it would have to be set dynamically per each file. In fact, in second case (with multiple linked css files in index.html) if I set relativeTo to ''src/lib/bootstrap/css', it would compile correctly and rebase paths correctly - but obviously, just for bootstrap and incorrectly for everything else. 

In that case, I don't understand how rewriting <link> elements to @import elements would change anything?

I guess what I'm asking here is, is there a way to provide a dynamic relativeTo to cleanCSS?

And a sidequestion - do 4.x change any of the behavior discussed here (other than using rebaseTo vs target/relativeTo)?

**Comment by sebastian-zarzycki-es** [comment]
Ok, scratch that, I can see that gulp-clean-css gets just one file after useref, so there's no way to get all the relativeTo paths at this stage. Which indeed, means, that the change has to be done on the useref level. I still could use an explanation why changing <link> into @import would help?

**Comment by sebastian-zarzycki-es** [comment]
This seems to work!

gulp.task('build', ['lib-resources'], function () {
  return gulp.src('src/index.html')
	.pipe(useref({}, lazypipe().pipe(function () {
		return gulpif('*.css', cleanCSS({ target: sourceDir }));
	})))	
    .pipe(uncache())
    .pipe(gulp.dest(targetDir));
});

**Comment by jakubpawlowicz** [comment]
Great - I didn't know you can do it that way too.

Re rewriting into `@import` - this is exactly what we've done manually by extracting `<link...` into `application.css`. Useref could do it in-memory and just use it from that point.

Regarding clean-css 4.x you need to replace `target` with `rebaseTo` and it should work fine.

**Comment by jakubpawlowicz** [comment]
See https://github.com/jakubpawlowicz/gulp-clean-css-v3-example (clean-css 3 example)
and https://github.com/jakubpawlowicz/gulp-clean-css-v4-example (clean-css 4 example)

**Comment by sebastian-zarzycki-es** [comment]
My example above stopped working in gulp-clean-css 3.0.3. I've changed target to rebaseTo, but it doesn't seem to make any difference. Everything compiles but the rebasing is wrong again.

**Comment by jakubpawlowicz** [comment]
@sebastian-zarzycki-es please see my comment above with a link to clean-css 4 example

**Comment by sebastian-zarzycki-es** [comment]
I saw it before. Your examples, again, rely on @import, rather than including straight from index.html. 

My example *did* work with clean-css 3. There was no change to gulp-useref or lazypipe since then, and only those could potentially affect the outcome. This means that clean-css 4 breaks this functionality. By your own writing "this means that rebasing URLs and import inlining is much simpler but may not be (YMMV) as powerful as in 3.x;". Any idea how to achieve the same functionality in 4.x, or is it just removed and not coming back?

Once again, my example - working - here. But it works, because it uses gulp-clean-css 2.3.2 which is the last version before they switched to clean-css 4.x. Once you switch that to newest ("*") in package.json, and update the gulpfile to use rebaseTo instead of target, it no longer works the same.

[rebase-example.zip](https://github.com/jakubpawlowicz/clean-css/files/795100/rebase-example.zip)

**Comment by xt0rted** [comment]
I'm not sure if this is related, but when using v4 I was getting the correct output on my dev machine and my test server, but on my two production servers the output was incorrect. The correct output should have been `url(/content/...)` but instead it was `url(content/...)`.

After some digging I found that there were 3 versions of node being used
5.7.0 - production server
6.2.0 - test server
6.9.5 - dev machine

The machines using 6.2+ were the ones generating the correct paths. After switching all of the servers to use 6.9.5 the issue went away and my paths are all `/content` now.

**Comment by jakubpawlowicz** [comment]
That's interesting. Could be a `path` issue fixed somewhere in between 5.7 and 6.2? Thanks @xt0rted for sharing!
